File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();

        if (root == null){
            return res;
        }
        
        Deque<TreeNode> traversal = new LinkedList<>();
        traversal.add(root);

        while(traversal.size() != 0){
            List<TreeNode> temp = new ArrayList<>(traversal);
            List<Integer> values = new ArrayList<>();

            temp.stream()
                .map(treeNode -> treeNode.val)
                .forEach(values::add);

            res.add(values);

            for (TreeNode i : temp){
                if (i.left != null){
                    traversal.add(i.left);
                }
                if (i.right != null){
                    traversal.add(i.right);
                }

                traversal.pollFirst();
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(Deque, self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque

class Codec:

    def serialize(self, root):
        if not root:
            return ""

        result = []
        queue = deque([root])

        while queue:
            node = queue.popleft()

            if node is None:
                result.append("N")
                continue

            result.append(str(node.val))

            queue.append(node.left)
            queue.append(node.right)

        return ",".join(result)

    def deserialize(self, data):
        if not data:
            return None

        values = data.split(",")
        root = TreeNode(int(values[0]))

        queue = deque([root])
        i = 1

        while queue:
            node = queue.popleft()

            # Left child
            if values[i] != "N":
                node.left = TreeNode(int(values[i]))
                queue.append(node.left)

            i += 1

            # Right child
            if values[i] != "N":
                node.right = TreeNode(int(values[i]))
                queue.append(node.right)

            i += 1

        return root